rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,267 @@
+
+
+### Observing pipelines
+
+Users should be able to observe and monitor individual pipelines.
+This means relevant metrics coming from the `internal_metrics` source must contain a `pipeline_id` tag refering to the pipeline `id` and a `pipeline_version` tag refering to the pipeline `version`.
+
+In Vector, the `Task` structure is what emits the events for `internal_metrics`.
+After [build the different pieces of the topology](https://github.com/timberio/vector/blob/v0.15.0/src/topology/builder.rs#L106), we've to update the [`Task::new`](https://github.com/timberio/vector/blob/v0.15.0/src/topology/builder.rs#L163) in order to accept an `Option<(PipelineId, PipelineVersion)>` so that when it emits the metrics events it can provide the information about the pipeline.
+
+This approach would extend the [RFC 2064](https://github.com/timberio/vector/blob/master/rfcs/2020-03-17-2064-event-driven-observability.md#collecting-uniform-context-data) by _just_ adding `pipeline_id` and `pipeline_version` to the context.
+
+When [spawning a transform](https://github.com/timberio/vector/blob/v0.15.0/src/topology/mod.rs#L574), adding the optional pipeline information to the span will populate the metrics.
+
+```rust
+let span = error_span!(
+    "transform",
+    component_kind = "transform",
+    component_name = %task.name(),
+    component_type = %task.typetag(),
+    pipeline_id = %task.pipeline_id(),
+    pipeline_version = %task.pipeline_version(),
+);
+```
+
+## Doc-level Proposal
+
+With this pipeline definition, the following `vector.toml`
+
+```toml
+[sources.docker]
+type = "docker_logs"
+...
+
+[transforms.nginx_filter]
+input = ["docker"]
+type = "remap"
+source = '''
+  # do something that filter nginx logs from docker logs
+'''
+
+[transforms.nginx]
+input = ["nginx_filter"]
+type = "remap"
+source = '''
+  # do some transformations with the nginx logs
+'''
+
+[transforms.apache_filter]
+input = ["docker"]
+type = "remap"
+source = '''
+  # do something that filter apache logs from docker logs
+'''
+
+[transforms.apache]
+input = ["apache_filter"]
+type = "remap"
+source = '''
+  # do some transformations with the apache logs
+'''
+
+[sinks.ouput]
+inputs = ["apache", "nginx"]
+type = "console"
+...
+```
+
+can be split into the following `pipelines/frontend.toml`
+
+```toml
+[transforms.nginx_filter]
+input = ["docker"]
+type = "remap"
+source = '''
+  # do something that filter nginx logs from docker logs
+'''
+
+[transforms.nginx]
+input = ["frontend.nginx_filter"]
+type = "remap"
+source = '''
+  # do some transformations with the nginx logs
+'''
+
+[transforms.apache_filter]
+input = ["docker"]
+type = "remap"
+source = '''
+  # do something that filter apache logs from docker logs
+'''
+
+[transforms.apache]
+input = ["frontend.apache_filter"]
+type = "remap"
+source = '''
+  # do some transformations with the apache logs
+'''
+```
+
+and `vector.toml`
```

Comment:
I'd like to see pipeline resource allocation here. Basically, settings that determine how global resources are made available to pipelines. Right now this is just sinks/sources, but it will also include new concepts like resources (being introduced in the [CSV enrichment RFC](https://github.com/timberio/vector/pull/8400)). To be specific:

- Defaults. Should new pipelines have access to all sources and sinks, specific ones, or none? How can admin change this?
- Individual pipeline allocations. Overriding defaults for each pipeline.